Hide windows until they are either rendered or a timeout expires

petemill opened the pull request:
Fix #8128

Waits for React to do its initial render of components, before firing an event to the browser process, which will show the window.

If the event is not received within 5 seconds (hard-coded, but in _config.js_), the window is shown (and will probably be a white screen).

Also:
- Separates parsing of APP_NEW_WINDOW action properties, and window creation / visibility logic.
- Moves window creation logic to windows API.
- Creates tests for existing APP_NEW_WINDOW action handler logic in windowsReducer
- Creates tests for new windows.createWindow logic
- Fixes test for downloadsReducer with not restoring a stub required by windowsReducer test

One complication is with macOS fullscreen windows for a couple reasons. 
1. macOS (or muon) will ignore the `show` property and open the window immediately if the window is fullscreen. This code counteracts that by removing the fullscreen property and then setting it once the window is ready (or the timeout happens)
2. When starting the app with many windows in the state to be opened, some of which are fullscreen and some of which are non-fullscreen on the same displays as a fullscreen window(s), the non fullscreen windows will open in the fullscreen space, which can be annoying for a user. I've gone between just having those windows show immediately and some more complex logic than what I have right now which attempted to open all non-fullscreen windows before opening fullscreen windows. The complex logic did not work because at that point, macOS would not put more than 1 window fullscreen per display even if the state was instructing it to. master currently supports that, and I didn't want to break it so right now is a balance (using a setTimeout of 100ms for fullscreen windows) between being able to open many fullscreen windows on app startup and preventing combined window spaces in most circumstances.

Thoughts for the future:
- Set a default (non-react) html content which displays that there has been an error, and provides tips for how to proceed.
- Remove the white background and put a gray one in (BrowserWindow has a background-color property)
- If fading in the windows is desired on some platforms where it is normal, we may need to [bring back muon's transparency property in BrowserWindow](https://github.com/brave/muon/issues/373), or use window handles.

## Testing

### Automated
`npm run unittest -- --grep "window API unit tests"`
`npm run unittest -- --grep "windowsReducer"`

### Manual
- Make sure windows open at app startup
 - Make sure windows loaded from state at app startup remember their locations and fullscreen status
- macOS / linux: when in a fullscreen window, ensure new windows opened are also fullscreen
- Try to have a window that errors (put `throw new Error()` in `entry.js`, it should open after 5 seconds
- Ideally if you have a profile that takes more than 5 seconds to open a window, you could observe that the window opens in a not-so-ready state after 5 seconds, but I have not been able to produce such a state.

Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request [as needed](https://github.com/brave/browser-laptop/wiki/Pull-request-process).

Reviewer Checklist:

Tests


- [x] Adequate test coverage exists to prevent regressions
- [x] Tests should be independent and work correctly when run individually or as a suite [ref](https://github.com/brave/browser-laptop/wiki/Code-Guidelines#test-dependencies)
- [x] New files have MPL2 license header

NejcZdovc commented:
@petemill did you maybe define wrong fix issue number?

petemill commented:
Yeah @NejcZdovc thanks, edited. Will edit commit message before merge.

bsclifton commented on app/browser/reducers/windowsReducer.js:
```diff
@@ -35,6 +35,8 @@ const appDispatcher = require('../../../js/dispatcher/appDispatcher')
 const isDarwin = platformUtil.isDarwin()
 const isWindows = platformUtil.isWindows()
 
+const TIMEOUT_WINDOW_SHOW_MS = 5000
```
I think this is fine here, but for future reference you may consider `js/constants/config.js` 😄

petemill replied:
moved to _config.js_

bsclifton commented on app/browser/reducers/windowsReducer.js:
```diff
@@ -245,7 +246,52 @@ const createWindow = (state, action) => {
+    // fullscreen once it is ready to be shown
+    // (browserOpts.fullscreen may already be set when loading from saved state,
+    // so this just sets it for other scenarios)
+    if (isDarwin && parentWindow && parentWindow.isFullScreen()) {
```
this logic (all of what is in this setImmediate) might be better to pull out to a discrete function.  This would help with unit testing and could help with readability. Something like `getCreateWindowArgs` (passing what's needed). Then whatever is returned is directly passed to the call `new BrowserWindow`

bsclifton replied:
Unit tests would then be a great way to capture the behavior (of setting fullscreen to false) and the tests themselves would act as documentation (if someone makes a change which breaks the test).  Could we get some unit tests for this?

petemill replied:
I have split all the window creation, showing, deferred showing, and timeout showing logic to a different place - _app/browser/windows.js_ which seemed an appropriate place which deals with calling functions on a muon BrowserWindow (and children). It's also the same place that the whenRendered action is handled, so that nicely puts all the logic in one place, including the temporary but ugly win.__blah properties - so they are at least encapsulated in one module.

I created tests for this new logic in _windowsTest.js_ handling:
- showing the window immediately
- showing the window if the timout expires
- showing the window on whenRendered firing for the window
- ensuring the window is fullscreen or maximized on each of those scenarios 

I also created tests for most of the existing action handler logic for APP_NEW_WINDOW in _windowReducerTest.js_, of which there were no existing tests, including tests for:
- creating the right tabs in the new window
- window position and size
- maximizing the window

bsclifton commented on test/unit/app/browser/reducers/windowsReducerTest.js:
```diff
@@ -20,36 +21,192 @@ describe('windowsReducer unit test', function () {
+      windowsReducer(state, action)
+      fakeTimers.tick(0)
+      // ensure the window api was asked to create the frame
+      const actualCreateWindowFrameArg = spy.args[0][3]
```
this is an interesting way to check the params- any reason you'd do this over something like `assert(spy.withArgs(args_here).calledOnce)`?

petemill replied:
@bsclifton In this case, sure that more expressive function could be used. I was just being consistent with the other tests here, which need to inspect the windowOptions object argument for property values, and not necessarily enforce every property on the object argument. If there's a nicer way to do that vs being tied to the argument order I'd love to learn!

petemill replied:
@bsclifton I updated the couple of tests I could using that syntax. It looks nicer, thanks. Only downside is the error messages aren't as expressive as using the chai.assert functions.

bsclifton commented on app/browser/reducers/windowsReducer.js:
```diff
@@ -325,11 +309,14 @@ const windowsReducer = (state, action, immutableAction) => {
     case appConstants.APP_WINDOW_READY:
       windows.windowReady(action.get('windowId'))
       break
+    case appConstants.APP_WINDOW_RENDERED:
```
Great having an event for this! 😄

bsclifton commented on app/browser/windows.js:
```diff
@@ -111,6 +120,26 @@ const updatePinnedTabs = (win) => {
   }
 }
 
+function showDeferredShowWindow (win) {
```
much better encapsulation this go-around 😄 👍

petemill commented:
Just noticed that multiple homepage tabs, as specified in https://community.brave.com/t/how-to-set-up-multiple-home-pages/9998 is not working in master (or this branch). I'll see if it's an easy fix we can bundle in this, or create the issue

bsclifton commented:
Found a problem- I grabbed latest (saw there was a rebase / force push after I last looked at code) and tried it out.  The sizes are not being saved for the windows ☹️  Here's a clip: basically the height for the window on the right is not saved
![not-saving-size](https://user-images.githubusercontent.com/4733304/33112185-19e46cb4-cf0f-11e7-9e9f-1a794cdc2536.gif)

bsclifton commented on app/browser/reducers/windowsReducer.js:
```diff
@@ -43,17 +39,43 @@ function isModal (browserOpts) {
+  const windowInfoState = immutableWindowState.get('windowInfo')
+  if (windowInfoState) {
+    browserOpts.width = firstDefinedValue(browserOpts.width, windowInfoState.get('width'))
+    browserOpts.height = firstDefinedValue(browserOpts.height, windowInfoState.get('windowInfo'))
```
Here's the bug causing the height issue 🐛  👀 !!

petemill replied:
oh good spot @bsclifton, I remember deleting that code 🤦‍♂️

bsclifton commented on app/browser/windows.js:
```diff
@@ -330,6 +372,108 @@ const api = {
+      // passed to BrowserWindow constructor
+    }
+    // let store know there's a new window
+    // so it can subscrive to state updates
```
s/subscrive/subscribe

bsclifton commented on test/unit/lib/fakeWindow.js:
```diff
@@ -3,17 +3,65 @@
-  getId () {
-    return this.id
+// cannot be a class since sinon has
+// trouble stubbing the constructtor for a class
```
s/constructtor/constructor

bsclifton commented on test/unit/lib/fakeElectron.js:
```diff
@@ -100,7 +88,16 @@ const fakeElectron = {
   }
 }
 
+// function getFakeDisplay
```
can this be removed?

bsclifton commented:
master 01e76241e53bbf4627e11aaee87411d5f40df9f4
0.21.x 57c20b682b6801a8b6342a368609210dbe11d565
0.20.x 0b27319147c8809e216a2f6e2feb555e480b5b58
0.19.x need some help merging this one, @petemill 😄

petemill commented:
@bsclifton I've got a successful merge on 0.19.x (pushed to [origin/0.19.x-hide-white-windows](https://github.com/brave/browser-laptop/commits/0.19.x-hide-white-windows)) - complication was that all the window creation logic had been moved and edited from appStore.js to windowReducer.js previous to this in #9963 and #10458, both of which were only merged to 0.20.x.  This could be a risk, but we've actually introduced more unit tests here, so up to you. If we do keep the 0.19.x merge, I would say incorporate the test plan from those PRs, but they both don't have one, so this actually would merge those fixes and create a test plan for them 🤷‍♂️ 😄 

I had to introduce all windowsReducerTests and windowsTests - and all the ones we created pass. However, there were two that were created before this PR that do not pass (and I removed them):
1. _windowsReducerTest.js_/`APP_WINDOW_UPDATED`/`updateDefault is true` is not passing because saving window dimensions is not working very well in 0.19.x. On the current 0.19.x, your [screencap above](https://github.com/brave/browser-laptop/pull/11764#issuecomment-346251228) doesn't work, i.e. the app only remembers one of the window positions, and layers all windows at that position, on top of eachother. This PR does not fix that, since it already had been fixed in master (in #10458). I removed that test.
2. _windowsTest.js_/`updatePinnedTabs` these tests were added as part of the #10531 fix, that is not in 0.19.x, so I removed them.

bsclifton commented:
Based on the above (and in the interest of reducing complexity), let's push this to 0.20.x 👍 

Thanks for taking the time to evaluate the merge 😄
